04.ls/ls.rb

```diff
@@ -1,19 +1,114 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'ffi-xattr'
+require 'etc'
+
 COLUMN_COUNT = 3
 SPACE = 1
+PERMISSION_TYPE = ['--x', '-w-', '-wx', 'r--', 'r-x', 'rw-', 'rwx'].freeze
 
 def main
   options = {}
   opt = OptionParser.new
+  opt.on('-r') { |v| v }
   opt.on('-a') { |v| v }
+  opt.on('-l') { |v| v }
   opt.parse!(ARGV, into: options)
 
   path =  "#{Dir.getwd}/"
   entries = Dir.entries(path).sort
-  filtered_entries = options.key?(:a) ? entries : entries.reject { |entry| entry.start_with?('.') }
-  entries_with_suffix = append_suffix_by_file_type(filtered_entries, path)
+
+  sorted_entries = options.key?(:r) ? entries.reverse : entries
+  filtered_entries = options.key?(:a) ? sorted_entries : sorted_entries.reject { |entry| entry.start_with?('.') }
+  options.key?(:l) ? show_long_format(filtered_entries, path) : show_short_format(filtered_entries, path)
+end
+
+def show_long_format(entries, path)
+  entries_long_format = []
+  total_blocks = 0
+
+  entries.each do |entry_name|
+    entry_details = File.lstat(entry_name)
+    total_blocks += entry_details.blocks
+    entries_long_format << load_details_each_entry(entry_name, entry_details, path)
+  end
+  print_long_data(entries_long_format, total_blocks)
+end
+
+def load_details_each_entry(entry_name, entry_info, path)
+  xattr = Xattr::Lib.list(path + entry_name, @no_follow = true)
+  entry_details_array = []
+
+  entry_details_array << (!xattr.empty? ? "#{export_permission(entry_info)}@" : "#{export_permission(entry_info)} ")
```

Comment:
```suggestion
  entry_details_array << (xattr.empty? ? "#{export_permission(entry_info)} " : "#{export_permission(entry_info)}@")
```

とした方が理解しやすいと思います。（なるべく否定の条件分岐を避ける）

あと`export_permission(entry_info)`が重複しているのでDRYにしたいです。